File: sections/28_dp_i_mixed_contest/problems/j_minimum_starting_energy/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, m = data[:2]
    values = data[2:]
    grid = [values[row * m:(row + 1) * m] for row in range(n)]
    infinity = 10**30
    need = [infinity] * m

    for row in range(n - 1, -1, -1):
        for column in range(m - 1, -1, -1):
            if row == n - 1 and column == m - 1:
                need[column] = max(0, -grid[row][column])
                continue
            after = infinity
            if row + 1 < n:
                after = min(after, need[column])
            if column + 1 < m:
                after = min(after, need[column + 1])
            need[column] = max(0, after - grid[row][column])

    print(need[0])


if __name__ == "__main__":
    main()

File: sections/28_dp_i_mixed_contest/problems/j_minimum_starting_energy/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    vector<vector<long long>> grid(n, vector<long long>(m));
    for (auto& row : grid)
        for (long long& value : row)
            cin >> value;

    const long long INF = 1LL << 62;
    vector<long long> need(m, INF);

    for (int row = n - 1; row >= 0; --row) {
        for (int column = m - 1; column >= 0; --column) {
            if (row == n - 1 && column == m - 1) {
                need[column] = max(0LL, -grid[row][column]);
                continue;
            }

            long long after = INF;
            if (row + 1 < n)
                after = min(after, need[column]);
            if (column + 1 < m)
                after = min(after, need[column + 1]);
            need[column] = max(0LL, after - grid[row][column]);
        }
    }

    cout << need[0] << '\n';
}
